### PR TITLE
Two lib effect dependencies

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: *(direct link to the issue)*

Two toolkit-neutral libraries that a later toolkit neutral lib-effects will depend on.

One for visiting effect or command parameters, which the macro system uses.

Another for the small language used by NumericTextCtrl for number formats, and for the
per-project format preferences.

QA:  Opening and using controls of almost any plain dialog in Audacity will sanity check this.

Also, check the working of changes of time format displays in the selection bar.  (And perhaps also frequency format changes in the spectral selection toolbar which is hidden by default.)

Also verify that the recently added seconds + milliseconds time format still works

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Review by vsverchinsky — COMMENTED** [review]
Artificial dependency on `lib-preferences` could be removed by further splitting `lib-command-parameters` into `lib-shuttle` and `lib-command-parameters`, and moving `ShuttlePrefs` class definition to `lib-preferences`

File: libraries/lib-command-parameters/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+set( SOURCES
+   Shuttle.cpp
+   Shuttle.h
+   ShuttlePrefs.cpp
+   ShuttlePrefs.h
+   WrappedType.cpp
+   WrappedType.h
+)
+set( LIBRARIES
+   lib-preferences-interface
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-command-parameters/CMakeLists.txt:17]: `lib-command-parameters` is the name of the new library.

I think I like this suggestion of a dependency inversion, though it is not needed for the present purpose of getting a `lib-effects`.

I will do this!

File: libraries/lib-command-parameters/CMakeLists.txt [hunk 2]
```diff
@@ -6,6 +6,8 @@ effect, or command
 set( SOURCES
    Shuttle.cpp
    Shuttle.h
+   ShuttleAutomation.cpp
+   ShuttleAutomation.h
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-command-parameters/CMakeLists.txt:10]: I see, there is no `lib-shuttle` yet but you think and even finer library subdivision is justified.

[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-command-parameters/CMakeLists.txt:10]: I don't like the term "Shuttle" and I think "ParameterVisitor" would be more descriptive.

File: src/widgets/NumericTextCtrl.cpp [hunk 3]
```diff
@@ -346,17 +346,15 @@ class NumericField
 class DigitInfo
 {
 public:
-   DigitInfo(int _field, int _index, int _pos, wxRect _box)
+   DigitInfo(int _field, int _index, int _pos)
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/widgets/NumericTextCtrl.cpp]: `wxRect` is a class not available when there is dependency on `wxBase` so for that reason alone I had to do something to get toolkit neutrality.  Besides, the NumericConverter object should only deal with conversions between numbers and strings of digits, and not involve geometry.

File: libraries/lib-command-parameters/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-command-parameters/CMakeLists.txt:10]: If libraries should be multiplied this way, I would name them `lib-effect-parameters` depending on `lib-command-parameters`.  Effects are a special case of commands, from the point of view of macro programming.

**Comment by Paul-Licameli** [comment]
Now that you have reviewed it, do you think there is applicability to export?  You may need to define something similar.

But maybe you will define something better and it can replace this and be reused.

If you examined WrappedType -- essentially it's just a variant.

**Comment by crsib** [comment]
I've refreshed my memory and I don't think I understand how exactly WrappedType is a variant.

**Comment by Paul-Licameli** [comment]
Actually now, rather than split Shuttle.cpp, I think I'll just move it down to lib-components

**Comment by Paul-Licameli** [comment]
@vsverchinsky please review again.

I changed my mind about some things and didn't always do what you or I wrote above.

I just eliminate class Shuttle and define its only subclass ShuttlePrefs without virtual functions.  This class is used only in ShuttleGui, but really should be considered a doomed class -- I have another monster branch of some ShuttleGui rewrites that eliminates it.

I don't try to fix the naming of classes after all.

File: libraries/lib-command-parameters/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (vsverchinsky) [reply libraries/lib-command-parameters/CMakeLists.txt:15]: Dependency on `lib-preferences` still remain

**Review by vsverchinsky — APPROVED** [review]
I still think that dependency on `lib-preferences` is quite odd, but if you're telling that `ShuttlePrefs` is to be removed I have no objections then

**Comment by Paul-Licameli** [comment]
Rebased onto recent master just to resolve small conflicts.

**Comment by Paul-Licameli** [comment]
Rebased again onto recent master with some conflict resolution;  make sure not to lose the recent addition of seconds + ms time format

**Comment by Paul-Licameli** [comment]
Rebased again to resolve minor conflicts.

**Comment by dozzzzer** [comment]
Testing this PR revealed that:

- The Time toolbar UI is broken in the macOS and Windows builds:

Windows: 
image
macOS: 
image

- Selection toolbar works just fine, also when changing time formats
- Spectral Selection toolbar also shows no issues
- All dialogs look fine, compared to master build

**Comment by Paul-Licameli** [comment]
I added a commit to fix the problem @dozzzzer reported.

**Review by vsverchinsky — APPROVED** [review]

**Comment by dozzzzer** [comment]
Verified on macOS Ventura and Windows 11

**Comment by Paul-Licameli** [comment]
> I've refreshed my memory and I don't think I understand how exactly WrappedType is a variant.

It's like a variant anyway.  It has one field that is like a discriminator for a union, though it does not literally use a `unioin` for the other members.  It's a single type that really holds some piece belonging to one of a short, closed-ended list of other types.  It also provides operators to interconvert whatever it is holding to any of the other types.

**Pull request merged**